### PR TITLE
Feature/698 new e2e tests

### LINKED ISSUE #698: Add missing e2e tests for new Features in 4.0
We introduced some new features in Release 4.0 that are not yet covered with e2e tests:

- [x] New Dialog "Create Comparison..."
- [x] New Management View for Comparisons
- [x] Dialog for Displaying Logs for Comparisons
- [x] Recalculating Comparisons

I think one simple scenario for each of the mentioned features would already be enough to get that basically covered..

See stories #650 , #692 and #604

### PR DESCRIPTION
**Pull request by eggerschwiler**
This pull request adds e2e tests for features that were newly added in the last release.

Acceptance criteria from Story:
We introduced some new features in Release 4.0 that are not yet covered with e2e tests:

- [x] New Dialog "Create Comparison..."
- [x] New Management View for Comparisons
- [x] Dialog for Displaying Logs for Comparisons
- [x] Recalculating Comparisons

I think one simple scenario for each of the mentioned features would already be enough to get that basically covered..

See stories #650 , #692 and #604 

fixes #698

### CODE REVIEW

File: scenarioo-client/app/manage/comparisons/createComparisonModal.html [hunk 1]
```diff
@@ -139,7 +139,7 @@ <h1 class="modal-title">Create Comparison</h1>
 
     <div class="modal-footer">
         <button class="btn btn-secondary" data-dismiss="modal" ng-click="vm.cancel()">Cancel</button>
-        <button class="btn btn-primary" ng-click="vm.createComparison()">Create</button>
+        <button class="btn btn-primary" ng-click="vm.createComparison()" id="createComparisonBtn">Create</button>
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/app/manage/comparisons/createComparisonModal.html]: Thanks for the hint. Updated the id.

File: scenarioo-client/test/e2e/pages/createComparisonDialog.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    async openTargetBranchSelectionDropdown() {
+        return element.all(by.partialLinkText('Branch:')).first().click();
+    }
+
+    async openTargetBuildSelectionDropdown() {
+        return element.all(by.partialLinkText('Build:')).first().click();
+    }
+
+    async chooseTargetBranch(branchName) {
```
Comment (danielsuter) [inline scenarioo-client/test/e2e/pages/createComparisonDialog.ts]: type parameters

File: scenarioo-client/test/e2e/specs/manage_comparison.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                await ComparisonsPage.goToPage();
+                await ComparisonsPage.assertComparisonStatus(0, 'SUCCESS');
+                await step('comparison page with successful comparison');
+
+                await ComparisonsPage.recalculateComparison(0);
+                await ComparisonsPage.assertComparisonStatus(0, 'PROCESSING');
+                await step('comparison in status PROCESSING');
+            });
+
+        scenario('Comparison details and log')
```
Comment (danielsuter) [inline scenarioo-client/test/e2e/specs/manage_comparison.ts]: This test is a change detector.

File: scenarioo-client/test/e2e/specs/manage_comparison.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        scenario('Filter comparisons')
+            .description('Comparisons can be filtered')
+            .it(async () => {
+                await ComparisonsPage.goToPage();
+                const comparisonsCount = ComparisonsPage.getNumberOfComparisons();
+                await step('comparisons page');
+
+                await ComparisonsPage.filterComparisons('To Projectstart');
+                await ComparisonsPage.assertNumberOfComparisons(5);
+                await step('filtered Comparisons for text \'To Projectstart\'');
```
Comment (danielsuter) [inline scenarioo-client/test/e2e/specs/manage_comparison.ts]: Use template strings so you don't have to escape

File: scenarioo-client/test/e2e/pages/comparisonDetailLogDialog.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+'use strict';
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonDetailLogDialog.ts]: I'm not sure. It is present in almost all of our ts-files, so I added it as well.

File: scenarioo-client/test/e2e/pages/comparisonsPage.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
+'use strict';
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonsPage.ts]: I'm not sure, but it is present in almost all of our ts-files, so at least it is consistent.

File: scenarioo-client/test/e2e/pages/createComparisonDialog.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
+'use strict';
+
+import {$, by, element} from 'protractor';
+import * as Utils from '../util';
+
+class CreateComparisonDialog {
+
+    async clickCreateComparisonLink() {
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/createComparisonDialog.ts]: createComparison() is already taken by the method function that actually creates the comparison. This function opens the Create Comparison Dialog. Renamed to openCreateComparison.

File: scenarioo-client/test/e2e/specs/manage_comparison.ts (hunk 3 above)
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/specs/manage_comparison.ts]: Sorry, the github code snippet doesn't clarify whether scenario "Recalculate comparison" or scenario "Comparison details and log" is meant.
What should be changed?

File: scenarioo-client/test/e2e/pages/comparisonDetailLogDialog.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+import {by, element} from 'protractor';
+
+class ComparisonDetailLogDialog {
+
+    async assertTitle(expectedTitle) {
+        return expect(element(by.tagName('h1')).getText()).toBe(expectedTitle);
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonDetailLogDialog.ts]: Thanks for the hint, I wasn't aware what the $ operator can do.

File: scenarioo-client/test/e2e/pages/comparisonDetailLogDialog.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
+'use strict';
+
+import {by, element} from 'protractor';
+
+class ComparisonDetailLogDialog {
+
+    async assertTitle(expectedTitle) {
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonDetailLogDialog.ts]: Thanks for the input. I'll keep a closer eye on type parameters in the future.

File: scenarioo-client/test/e2e/pages/comparisonDetailLogDialog.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    async assertComparedBuildDetails(expectedBranch, expectedBuild) {
+        return this.assertDetails('comparedBuildDetail', expectedBranch, expectedBuild);
+    }
+
+    async assertLogContains(expectedText) {
+        return expect(element(by.id('comparisonLog')).getText()).toContain(expectedText);
+    }
+
+    private async assertDetails(elementId: string, expectedBranch, expectedBuild) {
+        await expect(element(by.id(elementId)).getText()).toContain('Branch: ' + expectedBranch);
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonDetailLogDialog.ts]: Done.

File: scenarioo-client/test/e2e/pages/comparisonsPage.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+class ComparisonsPage {
+
+    private comparisonsTable = $('table.table-responsive');
+
+    async goToPage() {
+        return Utils.navigateToRoute('/manage?tab=comparisons');
+    }
+
+    async getNumberOfComparisons() {
+        const elements = this.comparisonsTable.all(by.css('tbody tr'));
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonsPage.ts]: Thanks for the hint, I wasn't aware what the $$ operator can do.

File: scenarioo-client/test/e2e/pages/comparisonsPage.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        return await elements.count();
+    }
+
+    async assertNumberOfComparisons(expectedCount) {
+        return Utils.assertNumberOfTableRows(this.comparisonsTable, expectedCount);
+    }
+
+    async clickRefreshLink() {
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonsPage.ts]: Done.

File: scenarioo-client/test/e2e/pages/comparisonsPage.ts [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        return Utils.clickElementById('refreshComparisons');
+    }
+
+    async clickResetButton() {
+        return Utils.clickElementById('resetComparisonsSearchField');
+    }
+
+    async filterComparisons(comparisonName) {
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonsPage.ts]: Done.

File: scenarioo-client/test/e2e/pages/comparisonsPage.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        const comparisonsSearchField = element(by.id('comparisonsSearchField'));
+        return comparisonsSearchField.sendKeys(comparisonName);
+    }
+
+    async assertComparisonStatus(rowIndex, status) {
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonsPage.ts]: Done

File: scenarioo-client/test/e2e/pages/comparisonsPage.ts [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        const elements = this.comparisonsTable.all(by.css('tbody tr'));
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonsPage.ts]: Done.

File: scenarioo-client/test/e2e/pages/comparisonsPage.ts [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        const row = elements.get(rowIndex);
+        const cells = row.all(by.tagName('td'));
+        return expect(cells.get(5).getText()).toBe(status);
+    }
+
+    async recalculateComparison(rowIndex) {
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonsPage.ts]: Done

File: scenarioo-client/test/e2e/pages/comparisonsPage.ts [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        const elements = this.comparisonsTable.all(by.css('tbody tr'));
+        const row = elements.get(rowIndex);
+        return row.element(by.linkText('Recalculate')).click();
+    }
+
+    async openComparisonDetails(rowIndex) {
+        const elements = this.comparisonsTable.all(by.css('tbody tr'));
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/comparisonsPage.ts]: Done

File: scenarioo-client/test/e2e/pages/createComparisonDialog.ts [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+        return Utils.clickElementById('createComparisonBtn');
+    }
+
+    async enterComparisonName(comparisonName) {
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/createComparisonDialog.ts]: Done

File: scenarioo-client/test/e2e/pages/createComparisonDialog.ts [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+        const comparisonNameField = element(by.id('comparisonName'));
+        return comparisonNameField.sendKeys(comparisonName);
+    }
+
+    async openTargetBranchSelectionDropdown() {
+        return element.all(by.partialLinkText('Branch:')).first().click();
+    }
+
+    async openTargetBuildSelectionDropdown() {
```
[the remark this answers is not in the document]
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/createComparisonDialog.ts]: Done

File: scenarioo-client/test/e2e/pages/createComparisonDialog.ts (hunk 2 above)
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/pages/createComparisonDialog.ts]: Done

File: scenarioo-client/test/e2e/specs/manage_comparison.ts (hunk 4 above)
Comment (eggerschwiler) [reply scenarioo-client/test/e2e/specs/manage_comparison.ts]: I tried to use template string `filtered Comparisons for text 'To Projectstart'` but that led to an npm compile error. It worked with "filtered Comparisons for text 'To Projectstart'".

**Pull request merged**